### DISCUSSION TITLE
sqlite3 binary in the image

### DISCUSSION THREAD

**Discussion by Hr46ph** [dani-garcia/vaultwarden#4244] [Q&A] [answered]
sqlite3 supports live database dumps right? I can't find the binary when I go into the container to dump the database.

I am running on kubernetes with persistent storage volumes. I can `kubectl exec -it -n vaultwarden pods/vaultwarden-0 -- /bin/sh` but I cannot find the sqlite3 binary.

The only other way I see is bringing vaultwarden down with `kubectl scale statefulset -n vaultwarden vaultwarden --replicas 0` and bringing up another container that has the sqlite3 command available, attach the volumes from vaultwarden and perform my backup. Seems overly complicated.

The admin interface has a backup button. It creates a file similar to sqlite3 backup. Im sure the command is hidden somewhere 😄 .

**Comment by zacknewman** [comment]
[`sqlite3` is bundled](https://github.com/dani-garcia/vaultwarden/blob/main/Cargo.toml#L82) in Vaultwarden as is relatively common for applications that use `sqlite3`. If you want to use it outside of the application, then you'll need to install it separately.

**Reply by Hr46ph** [reply]
No offense but please read again :).

I am looking for that command **inside** the container. kubectl exec -it takes me in and I cannot find the binary there.

I tried find `-type -name '*sq*'` among other things.

So if its there, where is it?

**Reply by stefan0xC** [reply]
There is no extra sqlite3 binary as the library is compiled and linked into the Vaultwarden binary. See my answer below for how the backup file is made.

**Comment by stefan0xC** [comment]
https://github.com/dani-garcia/vaultwarden/blob/890e668071cffe2833834348e19bbef3c061d014/src/db/mod.rs#L369-L386

**Reply by Hr46ph** [reply]
Thank you, so that doesn't seem like something I can work with, at least not in an easy way. I didn't realize it was done that way.

Basically I'm just looking for a way to make a reliable backup without bringing the application down. Since sqlite can make live dumps, I was hoping to schedule the command to run inside the container. The documentation is clear and I understand all that, but with kubernetes and most storage providers  (Longhorn in my case) I cannot mount a PV when its already bound to Vaultwarden pod (its read/write/once). It means I need to scale vaultwarden replicas=0, spin up a temporary container that has sqlite3 binary, give it the data volume to make the dump and bring that down again. Then I can scale vaultwarden up again. This seems like overly complex for something that should be really simple?

But  maybe I misunderstand something and there's easier way to do this that I am missing? There are backup tools like Velero and Longhorn itself can make snapshots and backups too. The issue is that those will backup the live database which might end up in an inconsistent state. Velero supports pre and post hooks, but I don't see how I can use that apart from stopping Vaultwarden so it flushes buffers to disk. Ideally, I would use that hook to run sqlite3 (there we have it again 😛 ).

In my mind, it would be so much easier if sqlite3 would be builtin but like I said, I might be missing something here.

Appreciate the help!

**Reply by BlackDex** [reply, marked as answer]
There are other ways with K8s.
As an example: https://github.com/preimmortal/sidecar-backup
And there are probably others too.

**Reply by Hr46ph** [reply]
Aah I see! Thanks! 😄 

I was trying to do that but as a separate pod.

**Comment by trexx** [comment]
There are a few ways to do this in K8s.

* Create a cronjob that mounts in the PV that Vaultwarden uses. Use a container image (or make one yourself) that has sqlite3 and run a backup or VACUUM INTO command on it.
* Create a cronjob that runs curl and sends a POST request to the /admin/config/backup_db so vaultwarden can create the backup for you. 

Then you can have a separate process like volsync, k8sup, or velero to backup the whole PVC.
